[47] 일반 사용자 주문 리스트

ohsuha opened the pull request:
- 상품명으로 주문 내역 검색

ohsuha committed:
[47] 일반 사용자 주문 리스트

- 상품명으로 주문 내역 검색

f-lab-kai-c commented on src/main/java/org/example/commerce_site/common/util/PageConverter.java:
```diff
@@ -8,7 +9,16 @@
+
 		int start = (int)pageable.getOffset();
+
+		if (start >= content.size()) {
+			return new PageImpl<>(Collections.emptyList(), pageable, content.size());
+		}
```
이건 어떤 경우를 처리하는거에요?

ohsuha replied:
입력받은 페이지가 전체 페이지보다 클 경우(예: 총 3페이지인데 6페이지 보여달라고 요청 받았을때) 예외가 발생해, empty리스트를 반환하도록 했습니다

f-lab-kai-c commented on src/main/java/org/example/commerce_site/infrastructure/order/CustomOrderRepositoryImpl.java:
```diff
@@ -0,0 +1,79 @@
+			builder.and(QProduct.product.name.contains(keyword));
+		}
+
+		List<OrderResponseDto.Get> orderList = queryFactory.select(Projections.constructor(OrderResponseDto.Get.class,
+				QOrder.order.id,
+				QOrder.order.totalAmount,
+				QOrder.order.status,
+				Projections.list(Projections.constructor(OrderDetailResponseDto.Get.class,
+					QOrderDetail.orderDetail.createdAt,
+					QOrderDetail.orderDetail.id,
+					QOrderDetail.orderDetail.productId,
+					QOrderDetail.orderDetail.quantity,
+					QOrderDetail.orderDetail.order.id,
+					QOrderDetail.orderDetail.unitPrice,
+					QProduct.product.name,
+					QShipment.shipment.status,
+					QShipment.shipment.createdAt,
+					QShipment.shipment.updatedAt
+				))
+			))
+			.from(QOrder.order)
+			.leftJoin(QOrderDetail.orderDetail).on(QOrder.order.id.eq(QOrderDetail.orderDetail.order.id))
+			.leftJoin(QProduct.product).on(QOrderDetail.orderDetail.productId.eq(QProduct.product.id))
+			.leftJoin(QShipment.shipment).on(QOrderDetail.orderDetail.id.eq(QShipment.shipment.orderDetail.id))
+			.where(QOrder.order.userId.eq(userId).and(builder))
+			.offset(pageable.getOffset())
+			.limit(pageable.getPageSize())
+			.fetch();
```
이 쿼리의 성능은 어느정도가 나올까요?

ohsuha replied:
EXPLAIN ANALYZE 를 돌려본 결과 검색어가 없는경우 0.246 검색어가 있는경우는 0.266 이 나왔습니다.
orderDetail 을 inner join 으로 바꾸면  양쪽 테이블에 조건에 맞는 데이터가 있을 때만 조인을 성사하기 때문에 불필요한 orderDetail 데이터를 가져오지 않아 조금 더 빨라진다고 해서 innerJoin 으로 변경하고,
상품명으로 검색하는 부분 products 테이블의 name 을 fullText 인덱스로 사용하려합니다

f-lab-kai-c commented on src/main/java/org/example/commerce_site/infrastructure/order/CustomOrderRepositoryImpl.java:
```diff
@@ -0,0 +1,79 @@
+		BooleanBuilder builder = new BooleanBuilder();
+
+		if (StringUtils.hasText(keyword)) {
+			builder.and(QProduct.product.name.contains(keyword));
```
이 조건을 위한 index를 설정한다고 어떻게 하면 될까요?

ohsuha replied:
fullText 인덱스로 설정하는 방법과 일반 like 인덱스로 설정하는 방법 두가지를 찾았습니다.
- like :  부분 문자열 검색(예: %apple%)에 대해서는 일반적으로 인덱스를 사용할 수 없다. 패턴이 문자열의 시작에서부터 일치해야한다. 
- FullText : 단어 단위로 검색하면 의미 있는 결과를 보다 쉽게 얻을 수 있다.  예를 들어, "apple"을 검색할 때 "apple"이라는 단어를 포함하는 모든 문서를 찾을 수 있다. FULLTEXT 검색은 같은 의미를 가진 다른 형태의 단어(예: "running", "run")도 처리할 수 있다. 

이런 이유로 FullText 인덱스로 설정하려고합니다.

ohsuha replied:
FullText 만으로는 iphone16 이라는 상품을 검색할때 'iph' 만으로 찾아지지 않아 like 와 FullText 를 병행하여 사용하도록 수정했습니다. 쿼리 성능은 인덱스를 사용하지 않을때는 0.246 에서 0.121 로 빨라졌습니다

f-lab-kai-c replied:
fulltext를 사용하면 어떻게 이런것들이 가능한걸까요?

ohsuha committed:
[47] 쿼리 조회 성능 향상

- product name 을 FullText 인덱스로 지정
- queryDSL 에서 fulltext MATCH... AGAINST 지원이 안되므로 sql native query 를 사용하도록 수정

ohsuha committed:
[41] rest template 을 connection pool 로 관리하기

f-lab-kai-c commented on src/main/java/org/example/commerce_site/common/util/PageConverter.java:
```diff
@@ -8,7 +9,16 @@
 
 public class PageConverter {
 	public static <T> Page<T> getPage(List<T> content, Pageable pageable) {
```
content 전체를 받고 일부를 잘라서 사용하는 것 같은데, 전체 row의 갯수가 엄청 많아지면 무슨일이 생길까요?

ohsuha replied:
굳이 필요없는 데이터까지 가져오므로 많이 느려질것같네요.. 쿼리에서 limit 을 주고 Total 값을 얻는 쿼리를 한번 더 돌리는 방법으로 수정 하겠습니다..!

ohsuha replied:
이부분 [48] 파트너회원 order list 브랜치에서 반영했습니다..!

ohsuha committed:
[41] rest template 에러 처리 수정

ohsuha committed:
Merge pull request #51 from f-lab-edu/feature/common/41-rest-tamplate-connect-pool

[41] rest template 을 connection pool 로 관리하기
